Add ability to ignore a command flag if default is nil

samwoods1 opened the pull request:
I've got the optional param: 
  t.add_flag do |flag|
    flag.name = '--load-path'
    flag.default = nil
    flag.message = 'Additional load path for Beaker'
    flag.override_env = 'BEAKER_LOAD_PATH'
  end

And I wanted to be able to have it not included at all if the default was nil.  Maybe there's a better way to indicate that it is optional?  Would we prefer an additional 'optional' param for flag?

samwoods1 committed:
(QA-2497) Add ability to ignore a command flag if required=false

samwoods1 commented:
Ok... So after working through this I kept running into various scenarios that didn't quite work, so I came up with some requirements:
- flag not required, default to 'not include', overridable by env_var
- flag not required, default value, overridable by env_var to different value
- flag not required, default value, overridable by env_var to 'not include'
- flag required, no default value, overridable by env_var
- flag required, default value, overridable by env_var

I think this approach fulfills all of the requirements:
- Add a new 'required' argument to add_flag that defaults to true. If set to true, everything behaves as it does currently.  If set to false then this would happen:
  - if the default is set to nil or empty string, or not set to anything then the flag will not be included in the command string unless overridden by an env_var
  - if the default is set to some value, but the env_var is set to empty string then the flag will not be included in the command string.

I'm working on updating the PR now.  Fixing unit tests.

er0ck commented:
> flag not required, default to 'not include', overridable by env_var

are you saying that you'd set {{:default}} to 'not include' and that string should somehow be 'magic' and make the flag disappear from the command string?

edit: oh, i think you are saying it's just by default not included in the command string.  but if you have an env_var for its value, it'd show up, followed by its value.
yeah.  i'm toying with it now.  i'll be sure to include these requirements.  thanks!

edit2: 
i think your requirements are good, but the API feels very complicated to me.  so complicated that i can't remember how it works, and i just read it.  we might just have to be verbose here, and make the API consistent, and lengthy, but discoverable and predictable.
e.g.: i'm toying with {{add_flag({:name, :default, :override_env, :boolean, :key_override_env}}
where we just have params that dictate if it's boolean and any env variable(s) that override the key/body of the flag.  i HATE my current naming of those new params, tho

er0ck commented:
added some labels just for kicks.  feel free to ignore yard docs and human docs, if it's too heavy, in this timeframe.
"Needs Human Docs" is for "human readable", markdown docs

er0ck commented:
for future reference... it might be advantageous to have the typos, etc in a separate (maint) commit, so if we have to back-out your other changes, we get to keep the speling fixes

er0ck commented:
👍 LGTM.  note, we are cutting some corners here, under the assumption that we need to ship this thing, and we're going to gut the API next release

zreichert commented on lib/rototiller/task/rototiller_task.rb:
```diff
@@ -46,12 +46,14 @@ def self.define_task(*args, &task_block)
       # @option args [String] :default The default value for the environment variable
       # @option args [String] :message A message describing the use of this variable
+      # @option args [Boolean] :required Indicates whether an error should be raised
+      # if the value is nil or empty string, vs not including the flag.
```
Should this be 'vs not manage the environment variable'?

samwoods1 replied:
No... It really only makes sense in the context of a CommandFlag, but I need it in EnvVar since it is used internally by CommandFlag and contains the logic to determine what the final value for the flag will be.  Since we would never expect a user to specify :required for just an EnvVar, and only expect CommandFlag to use it internally when creating an EnvVar would we want to remove this from the documentation?

samwoods1 commented:
@er0ck Just curious, what corners are we cutting?  If there's additional stuff required, let me know.

er0ck commented:
acceptance test would be great.  but i hesitate as the api will change so much, hopefully quickly.  then again, an acceptance test that tests the high level feature of being able to turn off an option and its value would be valuable now and later (v0.2.z)

i think i was mostly just trying (and failing) to state for posterity that this isn't really the best solution for the API but we've painted ourselves into a corner.  the RototillerTask api needs re-work before we move on to beaker tasks, etc.

samwoods1 commented:
Added acceptance tests, which uncovered some stuff, so also fixed some logic and now convert 'true' string to boolean (acceptance tests were assuming everything was a string)

er0ck commented on README.md:
```diff
@@ -48,6 +48,19 @@ Rototiller provides a Rake DSL addition called 'rototiller_task' which is a full
       end
     end
 
+    desc "do not include flag if value is nil or empty and not required"
```
nit: the word "value" is kinda ambiguous here

samwoods1 replied:
Yeah, I was trying to say that it is the final value after the default and override_env are evaluated.  What would be a more self explanatory word for that?  final value?

er0ck commented on acceptance/tests/rototillerTask/override_command_flag_with_env_var.rb:
```diff
@@ -105,4 +105,38 @@ def remove_reserved_keys(h)
   end
+
+  step 'Test flags that are not required' do
+    override = (0...10).map { ('a'..'z').to_a[rand(26)] }.join
```
we have {{random_string()}} in the acceptance lib
but it upcases.  that upcase should probably be removed.  one can upcase their own lazy self.
anyway, use that random_string helper here.

will require quick update to {{command_arguments_with_override.rb}}  (add .upcase to random_string)

er0ck commented on acceptance/tests/rototillerTask/override_command_flag_with_env_var.rb:
```diff
@@ -105,4 +105,38 @@ def remove_reserved_keys(h)
+
+        command_flags.each do |flag|
+          regex = /The CLI flag #{flag[:name]} has no value assigned and will not be included./
+          assert_match(regex, result.stdout, "The expected output from rototiller was not observed")
```
differentiate this assertion log message from the one above

er0ck commented on lib/rototiller/utilities/env_var.rb:
```diff
@@ -67,6 +73,11 @@ def message=(message)
+  def required=(required)
+    @required = required
+    reset
+  end
```
please comment wtf this is doing 😟

samwoods1 replied:
Added a comment above this block of variable assignment methods that are all doing the same thing.

er0ck replied:
🆗

er0ck commented on lib/rototiller/utilities/env_var.rb:
```diff
@@ -1,10 +1,10 @@
   include ColorText
 
-  attr_accessor :var, :message, :default
+  attr_accessor :var, :message, :default, :required
```
:nit these need yardoc comments, but they weren't there before, so i'm willing to kick it on down 🎢

zreichert commented on acceptance/tests/rototillerTask/override_command_flag_with_env_var.rb:
```diff
@@ -105,4 +107,38 @@ def remove_reserved_keys(h)
+    override = random_string
+    command_flags = [
+        {:name => '--not-required', :override_env => 'required_override', :required => false},
+        {:name => '--not-required-with-default', :override_env => override, :default => 'def_val', :required => false},
```
Sorry if I am being dense but in what circumstance does the :default => 'def_val' get used?  Followup question if I set override in the ENV does it use that value?

samwoods1 replied:
def_val would be used, except that I am setting the override to an empty string (line 132).
Yes, if you set override to something other than an empty string, then it will use that value instead.

zreichert replied:
@samwoods1 Thanks for the clarification, makes sense.

zreichert replied:
👍  LGTM
